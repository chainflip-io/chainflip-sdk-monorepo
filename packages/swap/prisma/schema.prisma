// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// `pnpm prisma format` after changes :)

generator client {
  provider        = "prisma-client-js"
  // https://github.com/prisma/prisma/issues/2584#issuecomment-646414549
  // https://github.com/prisma/prisma/issues/6603#issuecomment-849979849
  output          = "../node_modules/.prisma/client"
  previewFeatures = ["multiSchema"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
  schemas  = ["public", "private"]
}

enum Chain {
  Polkadot
  Ethereum
  Bitcoin

  @@schema("public")
}

enum Asset {
  FLIP
  USDC
  DOT
  ETH
  BTC

  @@schema("public")
}

model SwapDepositChannel {
  id                    BigInt   @id @default(autoincrement())
  uuid                  String   @unique @default(uuid())
  srcAsset              Asset
  depositAddress        String
  expectedDepositAmount Decimal  @db.Decimal(30, 0)
  destAsset             Asset
  destAddress           String
  expiryBlock           Int
  issuedBlock           Int
  createdAt             DateTime @default(now())
  swaps                 Swap[]

  @@index([depositAddress])
  @@schema("public")
}

model Swap {
  id                        BigInt              @id @default(autoincrement())
  /// the swap id on the state chain
  nativeId                  BigInt              @unique
  srcAsset                  Asset
  depositAmount             Decimal              @db.Decimal(30, 0)
  intermediateAmount        Decimal?             @db.Decimal(30, 0)
  egressAmount              Decimal?             @db.Decimal(30, 0)
  destAsset                 Asset
  destAddress               String
  depositReceivedAt         DateTime
  depositReceivedBlockIndex String
  swapExecutedAt            DateTime?
  swapExecutedBlockIndex    String?
  egressCompletedAt         DateTime?
  egressCompletedBlockIndex String?
  swapDepositChannel        SwapDepositChannel? @relation(fields: [swapDepositChannelId], references: [id])
  swapDepositChannelId      BigInt?
  txHash                    String?             @unique
  egress                    Egress?
  createdAt                 DateTime            @default(now())
  updatedAt                 DateTime            @default(now()) @updatedAt

  @@schema("public")
}

model Egress {
  id          BigInt     @id @default(autoincrement())
  /// the egress id on the state chain
  nativeId    BigInt
  /// the swap id on the state chain
  chain       Chain
  amount      Decimal    @db.Decimal(30, 0)
  timestamp   DateTime
  swap        Swap?      @relation(fields: [swapId], references: [id])
  swapId      BigInt?    @unique
  broadcast   Broadcast? @relation(fields: [broadcastId], references: [id])
  broadcastId BigInt?
  createdAt   DateTime   @default(now())

  @@unique([nativeId, chain])
  @@schema("public")
}

model Broadcast {
  id       BigInt   @id @default(autoincrement())
  /// the broadcast id on the state chain
  nativeId BigInt
  chain    Chain
  egresses Egress[]

  @@unique([nativeId, chain])
  @@schema("public")
}

model State {
  id     Int @id @default(autoincrement())
  height Int @default(0)

  @@schema("private")
}

model MarketMaker {
  id        Int    @id @default(autoincrement())
  name      String @unique
  publicKey String

  @@schema("private")
}

enum ThirdPartyProtocol {
  lifi
  squid

  @@schema("public")
}

model ThirdPartySwap {
  id            BigInt             @id @default(autoincrement())
  uuid          String             @unique
  protocol      ThirdPartyProtocol
  routeResponse Json
  txHash        String
  txLink        String
  createdAt     DateTime           @default(now())

  @@schema("public")
}
