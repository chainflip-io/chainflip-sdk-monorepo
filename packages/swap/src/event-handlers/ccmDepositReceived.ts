import { z } from 'zod';
import { encodedAddress } from './common';
import { u128, u64 } from '../parsers';
import type { EventHandlerArgs } from '.';

const ccmChannelMetadataArgs = z.object({
  message: z.string(),
  gasBudget: u128,
});

const ccmDepositMetadataArgs = z.object({
  channelMetadata: ccmChannelMetadataArgs,
});

const ccmDepositReceivedArgs = z.object({
  ccmId: u64,
  principalSwapId: u64.nullable().optional(),
  gasSwapId: u64.nullable().optional(),
  depositAmount: u128,
  destinationAddress: encodedAddress,
  depositMetadata: ccmDepositMetadataArgs,
});

export type CcmDepositReceivedArgs = z.input<typeof ccmDepositReceivedArgs>;

export default async function ccmDepositReceived({
  prisma,
  event,
}: EventHandlerArgs) {
  const { principalSwapId, depositMetadata } = ccmDepositReceivedArgs.parse(
    event.args,
  );

  if (principalSwapId) {
    prisma.swap.update({
      where: {
        nativeId: principalSwapId,
      },
      data: {
        ccmGasBudget: depositMetadata.channelMetadata.gasBudget.toString(),
        ccmMessage: depositMetadata.channelMetadata.message,
      },
    });
  }
}
