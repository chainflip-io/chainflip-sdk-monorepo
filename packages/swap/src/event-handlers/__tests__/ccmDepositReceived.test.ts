import prisma from '@/swap/client';
import ccmDepositReceived, { CcmDepositReceivedArgs } from '../ccmDepositReceived';

describe(ccmDepositReceived, () => {
  beforeEach(async () => {
    await prisma.$queryRaw`TRUNCATE TABLE "SwapDepositChannel", "Swap", "SwapRequest" CASCADE`;
  });

  it('updates the CCM fields', async () => {
    const block = {
      timestamp: new Date().toISOString(),
      height: 1000,
      specId: 'test@150',
      hash: '0x123',
    };
    const swapRequest = await prisma.swapRequest.create({
      data: {
        nativeId: 1n,
        originType: 'VAULT',
        requestType: 'LEGACY_SWAP',
        srcAsset: 'Eth',
        destAsset: 'ArbEth',
        destAddress: '0xa56A6be23b6Cf39D9448FF6e897C29c41c8fbDFF',
        depositAmount: '1000000000000000000',
        swapRequestedAt: new Date('2024-08-23 13:10:06.000+00'),
        ccmGasBudget: null,
        ccmDepositReceivedBlockIndex: null,
        ccmMessage: null,
      },
    });

    const args: CcmDepositReceivedArgs = {
      principalSwapId: swapRequest.nativeId.toString(),
      destinationAddress: {
        __kind: 'Arb',
        value: '0xa56A6be23b6Cf39D9448FF6e897C29c41c8fbDFF',
      },
      ccmId: '1',
      depositAmount: '1000000000000000000',
      depositMetadata: {
        channelMetadata: {
          message: '0xcafebabe',
          gasBudget: '65000',
        },
      },
    };

    await ccmDepositReceived({
      prisma,
      block,
      event: {
        args,
        name: 'Swapping.CcmDepositReceived',
        indexInBlock: 6,
      },
    });

    expect(
      await prisma.swapRequest.findUniqueOrThrow({ where: { id: swapRequest.id } }),
    ).toMatchSnapshot({
      id: expect.any(BigInt),
    });
  });
});
