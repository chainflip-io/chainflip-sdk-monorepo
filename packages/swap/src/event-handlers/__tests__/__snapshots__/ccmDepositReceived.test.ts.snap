// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ccmDepositReceived updates the CCM fields 1`] = `
{
  "ccmDepositReceivedBlockIndex": "1000-6",
  "ccmGasBudget": "65000",
  "ccmMessage": "0xcafebabe",
  "completedAt": null,
  "completedBlockIndex": null,
  "depositAmount": "1000000000000000000",
  "depositBoostedAt": null,
  "depositBoostedBlockIndex": null,
  "depositReceivedAt": null,
  "depositReceivedBlockIndex": null,
  "depositTransactionRef": null,
  "destAddress": "0xa56A6be23b6Cf39D9448FF6e897C29c41c8fbDFF",
  "destAsset": "ArbEth",
  "effectiveBoostFeeBps": null,
  "egressId": null,
  "id": Any<BigInt>,
  "nativeId": 1n,
  "originType": "VAULT",
  "prewitnessedDepositId": null,
  "refundEgressId": null,
  "requestType": "LEGACY_SWAP",
  "srcAddress": null,
  "srcAsset": "Eth",
  "swapDepositChannelId": null,
  "swapRequestedAt": 2024-08-23T13:10:06.000Z,
}
`;
